### ISSUE TITLE
How do you implement the multi gpu distribution training part?

### ISSUE THREAD

**Issue by gaopengcuhk** [facebookresearch/XLM#153]
I am following the guidance of pytorch distribution training guidline which can be find here.

https://pytorch.org/tutorials/intermediate/dist_tuto.html

It seems that all you need is to split data-set and use all reduce operator to aggregate gradients over all gpu in the BP.

I have reading through your code careful and do not find where you implement this code. Can you point out where you split the data-set across GPU and aggregate all gradients.

**Comment by gaopengcuhk** [comment]
When I extend one gpu to n gpu, do I need to linearly scale the learning rate for BERT pretraining?

**Comment by glample** [comment]
Most of the GPU code is there: https://github.com/facebookresearch/XLM/blob/master/src/slurm.py

The dataset is not split across GPUs. Each GPU process loads the dataset, and trains on it. At the end of the first epoch (if the epoch size is the dataset size), you have basically seen your training set N times where N is the number of GPUs. This is different than how it's done in ImageNet for instance. Also, you do not need to scale the learning rate with Adam, but you need with other optimizers like SGD. Note that with Adam you can still increase it a bit if you have many GPUs, but it's fine to keep it the same.

**Comment by gaopengcuhk** [comment]
Thank you so much for your response.
I can understand the dataset loader part. However, as implied  in the tutorial, during BP  all-reduce will be used for gradient aggregation over all GPUs using the following code. 

def average_gradients(model):
    size = float(dist.get_world_size())
    for param in model.parameters():
        dist.all_reduce(param.grad.data, op=dist.reduce_op.SUM)
        param.grad.data /= size

It seems like your codebase do not contain this.

**Comment by glample** [comment]
With the distributed wrapper on top of the model, the gradient synchronization is done automatically whenever you call backward. You never have to manually do the `all_reduce`.

**Comment by gaopengcuhk** [comment]
Thank you for your response.

It seems this module implement all_reduce operator. 

 setattr(self, name, nn.parallel.DistributedDataParallel(getattr(self, name), device_ids=[params.local_rank], output_device=params.local_rank, broadcast_buffers=True))
